sphinx/ext/autodoc.py

```diff
@@ -171,13 +171,15 @@ def __init__(self, names):
             #     set(['a', 'd'])
             self.base_packages.add(n.split('.')[0])
         self.mocked_modules = []  # type: List[str]
-        self.orig_meta_path = sys.meta_path
         # enable hook by adding itself to meta_path
         sys.meta_path = sys.meta_path + [self]
 
     def disable(self):
         # restore original meta_path to disable import hook
-        sys.meta_path = self.orig_meta_path
+        for idx, importer in enumerate(sys.meta_path):
+            if importer is self:
+                del sys.meta_path[idx]
+                break
```

Comment:
@shimizukawa , the problem is not in a new list and restoring `self.orig_meta_path` is exactly the cause of the issue.

Here is the situation (you can check all the details in the description of the issue):
- some packages were imported so in `sys.meta_path` there will be `[importerA, importerB]` which is `self.orig_meta_path`
- then `MockImporter` was added to `sys.meta_path`, so now it contains `[importerA, importerB, MockImporter]`
- then something else was imported so in `sys.meta_path` there is `[importerA, importerB, MockImporter, importerC]`
 - then sphinx cleans up its `MockImporter` by putting `self.orig_meta_path` into `sys.meta_path` and `importerC` is lost!

So what fix does, it removes just 'MockImporter' and does it by comparison of its `id`.

Does it make sense? Let me know which parts of the issue is unclear I will elaborate more.